packages/eslint-plugin/src/rules/await-thenable.ts

```diff
@@ -26,9 +26,9 @@ export default util.createRule({
 
     return {
       AwaitExpression(node) {
-        const originalNode = parserServices.esTreeNodeToTSNodeMap.get(
-          node,
-        ) as ts.AwaitExpression;
+        const originalNode = parserServices.esTreeNodeToTSNodeMap.get<
+          ts.AwaitExpression
+        >(node);
```

Comment:
Is this explicit cast still necessary? Is it possible to more strongly type this map? Should we instead provide an accessor function that is more type-safe (by converting the type of the passed-in node to the type of the returned node)?

Reply:
potentially!
I'd have to defer to someone like @uniqueiniquity / @armano2 / @JamesHenry on that, because I don't know enough about the mapping between the node types.

Is it exactly a relationship that we can express by hand? ie can we do
```TS
interface ParserWeakMap<TKey, TValueBase> {
  get(key: TSESTree.AwaitExpression): ts.AwaitExpression;
  get(key: TSESTree.X): ts.Y;
  // ...
  has(key: any): boolean;
}
```

Reply:
(either way probs should explore this in a separate PR to reduce diff noise)

Reply:
Yeah I believe we could define a type for the map that explicitly enumerates the expected TS type for each input ES type (and vice versa for the other one). But I would agree that doing it in a separate PR would be good for noise reasons.
I'd be a little worried about it getting out of sync with changes to typescript-estree/convert.ts but maybe there could be a commit hook or test to prevent that somehow.